src/__tests__/with-theme.js

```diff
@@ -0,0 +1,101 @@
+/* eslint func-style:0 */
+import React, {Component} from 'react'
+import {render, mount} from 'enzyme'
+
+import withTheme from '../with-theme'
+import ThemeProvider, {CHANNEL} from '../theme-provider'
+import {PropTypes} from '../react-compat'
+
+const getMockedContext = unsubscribe => ({
+  [CHANNEL]: {
+    getState: () => {},
+    setState: () => {},
+    subscribe: () => unsubscribe,
+  },
+})
+
+test('renders a non-glamorous component with theme', () => {
+  const CompWithTheme = withTheme(({theme: {padding}}) => (
+    <div style={{padding}} />
+  ))
+  expect(
+    render(
+      <ThemeProvider theme={{padding: '10px'}}>
+        <CompWithTheme />
+      </ThemeProvider>,
+    ),
+  ).toMatchSnapshot()
+})
+
+test('theme properties updates get propagated down the tree', () => {
+  class Parent extends Component {
+    state = {
+      padding: 10,
+    }
+
+    render() {
+      return (
+        <ThemeProvider theme={{padding: this.state.padding}}>
+          <Child />
+        </ThemeProvider>
+      )
+    }
+  }
+
+  const Child = withTheme(({theme: {padding}}) => <div style={{padding}} />)
+  const wrapper = mount(<Parent />)
+  expect(wrapper).toMatchSnapshot(`with theme prop of padding 10px`)
+  wrapper.setState({padding: 20})
+  expect(wrapper).toMatchSnapshot(`with theme prop of padding 20px`)
+})
+
+test('works properly with classes', () => {
+  /* eslint-disable react/prefer-stateless-function */
+  class Child extends Component {
+    render() {
+      const {theme: {padding}} = this.props
+      return <div style={{padding}} />
+    }
+  }
+
+  Child.propTypes = {
+    theme: PropTypes.object,
+  }
+
+  const ChildWithTheme = withTheme(Child)
+
+  class Parent extends Component {
+    state = {
+      padding: 10,
+    }
+
+    render() {
+      return (
+        <ThemeProvider theme={{padding: this.state.padding}}>
+          <ChildWithTheme />
+        </ThemeProvider>
+      )
+    }
+  }
+
+  const wrapper = mount(<Parent />)
+  expect(wrapper).toMatchSnapshot(`with theme prop of padding 10px`)
+})
+
+test('pass through when no theme provider found up tree', () => {
+  const Child = withTheme(() => <div />)
+  const Parent = () => <Child />
+  const wrapper = mount(<Parent />)
+  expect(wrapper).toMatchSnapshot(`without theme provider`)
```

Comment:
Could we add an assertion here that `console.warn` was called? We'd need to mock `console.warn`:

```javascript
const originalWarn = console.warn
console.warn = jest.fn()
// your test code you have already
expect(console.warn).toHaveBeenCalledTimes(1)
expect(console.warn). toHaveBeenCalledWith('Expected ThemeProvider but none found!')
```

And actually, if we could make that warning a little more helpful that'd be great. Like providing `ComponentToTheme .displayName || ComponentToTheme .name` and saying it's coming from glamorous. So like:

```javascript
// eslint-ignore-next-line max-len
console.warn(`glamorous warning: Expected component called ${ComponentToTheme .displayName || ComponentToTheme .name} which uses withTheme to be within a ThemeProvider but none was found.`)
```

I don't mind being verbose about this :smile: